base_layer/core/src/proof_of_work/monero_rx.rs

```diff
@@ -22,97 +22,323 @@
+        tari_header.pow.target_difficulty = Difficulty::from(1);
+    }
+
+    /// This function adds in the tari header to the coinbase
+    pub fn adjust_mm_tag(&mut self, tari_header: &BlockHeader) {
+        self.coinbase_tx.prefix.extra.0.clear();
+        self.coinbase_tx
+            .prefix
+            .extra
+            .0
+            .push(merged_mining_subfield(tari_header));
+    }
 }
 
 /// Calculate the difficulty attained for the given block deserialized the Monero header from the provided header
 pub fn monero_difficulty(header: &BlockHeader) -> Difficulty {
+    monero_difficulty_with_hash(header).0
+}
+
+pub fn monero_difficulty_with_hash(header: &BlockHeader) -> (Difficulty, Vec<u8>) {
     match monero_difficulty_calculation(header) {
-        Ok(v) => v,
-        Err(_) => 1.into(), // todo this needs to change to 0 when merge mine is implemented
+        Ok(d) => (d.0, d.1),
+        Err(_) => (1.into(), Hash::null_hash().0.to_vec()),
     }
 }
 
+pub fn merged_mining_subfield(header: &BlockHeader) -> SubField {
+    let tag = merged_mining_tag_hash(header);
+    let depth = 0;
+    SubField::MergeMining(VarInt(depth), tag)
+}
+
+fn merged_mining_transaction_hash(header: &BlockHeader) -> Hash {
+    let mut merge_mining_prehash: Vec<u8> = header.version.to_le_bytes().to_vec();
+    merge_mining_prehash.append(&mut header.height.to_le_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.prev_hash.as_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.timestamp.as_u64().to_le_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.output_mr.as_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.range_proof_mr.as_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.kernel_mr.as_bytes().to_vec());
+    merge_mining_prehash.append(&mut header.total_kernel_offset.as_bytes().to_vec());
+    Hash::hash(merge_mining_prehash.as_slice())
+}
+
+fn merged_mining_tag_hash(header: &BlockHeader) -> Hash {
+    merged_mining_transaction_hash(header)
+}
+
 /// Internal function to calculate the difficulty attained for the given block Deserialized the Monero header from the
 /// provided header
-fn monero_difficulty_calculation(header: &BlockHeader) -> Result<Difficulty, MergeMineError> {
-    let monero = MoneroData::new(header)?;
-    verify_header(&header, &monero)?;
+fn monero_difficulty_calculation(header: &BlockHeader) -> Result<(Difficulty, Vec<u8>), MergeMineError> {
     #[cfg(feature = "monero_merge_mining")]
     {
+        println!("Deserializing monero header");
+        let monero = MoneroData::new(header)?;
+        println!("Verifying header");
+        verify_header(&header, &monero)?;
         let flags = RandomXFlag::get_recommended_flags();
-        let key = monero.key.clone();
+        // we need db access, this should be checked as part of the concensus code
+        // check_randomx_key(&monero)?;
```

Comment:
This check will need db access, we need to port this part to the statefull validation

Reply:
Think this has been addressed in #2061 as part of the block validation code?